### ISSUE TITLE
Bug in deployment: Instance is Stuck in the Initialization Loop [Amazon Linux 2]

### ISSUE THREAD

**Issue by serebrov** [aws/elastic-beanstalk-roadmap#75]
There is an issue with deployment on the new platform: sometimes the "Serial" field in the `/opt/elasticbeanstalk/deployment/app_version_manifest.json` is a string instead of number, for example:

```
cat /opt/elasticbeanstalk/deployment/app_version_manifest.json
{"RuntimeSources":{"backend":{"app-200716_221814":{"s3url":""}}},"DeploymentId":21,"Serial":"25"}
```

This causes an error in the `platform-engine -command=userdata-exec` command execution, and we can see the error in the `/var/log/eb-engine.log`:

```

2020/07/15 11:54:24.419932 [INFO] Downloading: bucket: elasticbeanstalk-us-east-1-280..., object: /resources/environments/e-ybspftm2u5/_runtime/versions/manifest_159...
2020/07/15 11:54:24.432921 [INFO] Download successful 97bytes downloaded
2020/07/15 11:54:24.432934 [INFO] Trying to read and parse version manifest...
2020/07/15 11:54:24.433089 [ERROR] failed to parse JSON file /opt/elasticbeanstalk/deployment/app_version_manifest.json with error: json: cannot unmarshal string into Go struct field AppVersionManifest.Serial of type uint64
... 
(process repeats endlessly)
```

I didn't find a reason for the "Serail" to be string sometimes, the way I fix this at the moment is to just run `eb deploy` manually to push the new version to Elastic Beanstalk and, usually, the `app_version_manifest.json` is updated with an integer value and the initialization proceeds.

I am using new Elastic Beanstalk platform (Amazon Linux 2, python) for around a month and this is the only critical issue that I see.
It is also relevant for at least Ruby and PHP platforms too, see links below.

Related Links:
* [Server stuck in a loop (eb-engine.log) ERROR failed to parse JSON file](https://forums.aws.amazon.com/thread.jspa?messageID=947760#947760)
* [AWS Elastic Beanstalk won't deploy my Rails app even once](https://stackoverflow.com/questions/62184145/aws-elastic-beanstalk-wont-deploy-my-rails-app-even-once)

**Comment by jwtrhs** [comment]
Have also experienced the same issue. I'm using the ElasticBeanstalk docker platform, Amazon Linux 2 on version 3.0.3 with deploys coming from CodePipeline/CodeBuild.

Rebuilding the beanstalk environment seems to have fixed it for now.

**Comment by mtdewulf** [comment]
Same issue on platform: Node.js 12 running on 64bit Amazon Linux 2/5.1.0

**Comment by markotitel** [comment]
Looks like this happens when existing app version is deployed.

**Comment by mtdewulf** [comment]
> Looks like this happens when existing app version is deployed.

In my case, I had:

* Failed deploy (new eb extension command failed)
* Terminated ec2 instance to get fresh box
* Started getting this error

**Comment by janost** [comment]
I also experienced a similar issue with PHP 7.4 on AL2. EB extensions were also randomly failing with weird error messages, so I ended up rewriting the same functionality we had in `.ebextensions` as platform hooks and the problem went away, so I guess the issue might be somewhere around handling `.ebextensions`.

**Comment by mtdewulf** [comment]
I reverted my eb extension changes to the last working version and still can't get a successful deploy.

**Comment by serebrov** [comment]
One more manual fix suggested in comments on StackOverflow:

>  If you look in the logs, you'll see EB is downloading that file from S3 (2020/07/14 22:58:41.937541 
> [INFO] Downloading: bucket: elasticbeanstalk-us-east-1-603873217483, 
> object: /resources/environments/e-k9bptzbdbu/_runtime/versions/manifest_1594766763384) , where the Serial is a string.
> Change it to an Int there, and it works.
> Obviously that is not a long term fix, but keep an eye on the aws thread mentioned above because it's definitely an AWS bug 
> – Cody Swann Jul 14 at 23:19

Releated: [the AWS thread also mentioned in comments](https://forums.aws.amazon.com/thread.jspa?threadID=206033)

**Comment by mtdewulf** [comment]
I've tried editing the manifest files in S3, but unfortunately run into a different issue.  In my case, the last "good" deploy was actually not good -- it has invalid `.ebextensions` and fails on fresh instances.  When I fix the manifest in S3 to be a number, EB tries to deploy that last, bad version.

I noticed the manifest also lets you pick the app bundle, so I updated that field to a bundle that does not have the offending `.ebextensions`.  Unfortunately, though `/var/app/staging` has the correct code, it appears EB is still trying to execute invalid eb extensions and is failing to deploy.  These are somehow persisted in JSON files located in `/var/lib/cfn-init/`.
